contract/src/main.rs

```diff
@@ -799,12 +868,31 @@ pub extern "C" fn approve() {
         Some(operator),
     );
 
+    let events_mode = EventsMode::try_from(get_stored_value_with_user_errors::<u8>(
+        crate::constants::EVENTS_MODE,
+        NFTCoreError::MissingEventsMode,
+        NFTCoreError::InvalidEventsMode,
+    ))
+    .unwrap_or_revert();
+
+
     // Emit Approval event.
-    casper_event_standard::emit(Approval::new(token_owner_key, operator, token_identifier));
+    match events_mode {
+        EventsMode::NoEvents => {}
+        EventsMode::CEP47 => {
+            let cep47_event = CEP47Event::Approve {
+                owner: token_owner_key,
+                spender: operator,
+                token_id: token_identifier
+            };
+            record_cep47_event_dictionary(&cep47_event)
+        },
+        EventsMode::CES => {
+            casper_event_standard::emit(Approval::new(token_owner_key, operator, token_identifier))
+        },
+    };
 }
 
-// This is an extremely gas intensive operation. DO NOT invoke this
-// entrypoint unless absoluteness necessary
 // Approves the specified operator for transfer token_owner's tokens.
 #[no_mangle]
 pub extern "C" fn set_approval_for_all() {
```

Comment:
At the end of this function, a CES event is emitted always, regardless of the event mode in the contract.

Reply:
Good catch! this means there is no negative case testing (set no events or cep47 and test no CES event has been emited and Vice versa )

Reply:
Good catch, I shall add those tests

Reply:
Confirmed with tests, that the approval event is not emitted if NoEvents is toggled on